test is failed with Union type?

```
...
metashape/tests/test_name.py::test_generics FAILED                       [ 20%]
metashape/tests/test_name.py::test_Literal PASSED                        [ 21%]
metashape/tests/test_name.py::test_Union FAILED                          [ 22%]
metashape/tests/test_name.py::test_NewType PASSED                        [ 23%]
...

=================================== FAILURES ===================================
________________________________ test_generics _________________________________

    def test_generics():
        import typing as t
    
        IntList = t.List[int]
        StrTuple1 = t.Tuple[str]
        StrTupleN = t.Tuple[str, ...]
        Commands = t.Dict[str, t.Callable[..., t.Any]]
    
        class Person:
            pass
    
        PersonDict = t.Dict[str, Person]
    
>       assert callFUT(IntList) == "IntList"
E       AssertionError: assert 'List' == 'IntList'
E         - IntList
E         + List

metashape/tests/test_name.py:45: AssertionError
__________________________________ test_Union __________________________________

    def test_Union():
        from typing import Union
    
        class X:
            pass
    
        class Y:
            pass
    
        XorY = Union[X, Y]
        XorY.__name__ = "XorY"
>       assert callFUT(XorY) == "XorY"
E       AssertionError: assert 'Union' == 'XorY'
E         - XorY
E         + Union

metashape/tests/test_name.py:74: AssertionError
=========================== short test summary info ============================
FAILED metashape/tests/test_name.py::test_generics - AssertionError: assert '...
FAILED metashape/tests/test_name.py::test_Union - AssertionError: assert 'Uni...
========================= 2 failed, 88 passed in 0.67s =========================
```

--- a/metashape/name.py
+++ b/metashape/name.py
@@ -122,18 +122,13 @@ class NameGuesser:
         return guessed
 
     def _guess(self, typ: t.Type[t.Any]) -> str:
-        name = self.resolver.resolve_maybe(typ)
-        if name is not None:
-            return name
-
-        alias = self._aliases.get(typ)
-        if alias is not None:
-            return alias
-
-        # for generics
-        origin = typ.__origin__
-        joiner = self._joiners.get(origin) or self._joiners[t.Any]
-        return joiner(origin, typ.__args__)
+        if not hasattr(typ, "__origin__"):
+            return self.resolver.resolve_maybe(typ) or self._aliases.get(typ)
+        else:
+            # for generics
+            origin = typ.__origin__
+            joiner = self._joiners.get(origin) or self._joiners[t.Any]
+            return joiner(origin, typ.__args__)
 
     def _join_name_for_default(
         self, origin: t.Type[t.Any], args: t.Tuple[t.Type[t.Any], ...],

--- a/metashape/name.py
+++ b/metashape/name.py
@@ -112,6 +112,10 @@ class NameGuesser:
 
         return weakref.WeakKeyDictionary()
 
+    def register(self, name: str, typ: t.Type[t.Any]) -> None:
+        self.resolver.register(name, typ)
+        self._cache[self.resolver._get_key(typ)] = name
+    
     def guess(self, typ: t.Type[t.Any]) -> str:
         k = self.resolver._get_key(typ)
         cached = self._cache.get(k)
@@ -173,7 +177,8 @@ _guesser = NameGuesser(_resolver)
 def NewNamedType(
     name: str, typ: TypeT, *, _resolver: NameResolver = _resolver
 ) -> TypeT:
-    _resolver.register(name, typ)
+    # TODO(podhmo): This is hack. (more complete implementation is https://github.com/lovasoa/marshmallow_dataclass)
+    _guesser.register(name, typ)
     return typ
 
 

